org.eclipse.buildship.core/src/main/java/org/eclipse/buildship/core/workspace/internal/DerivedResourcesUpdater.java

```diff
@@ -50,35 +70,78 @@ private void update(IProgressMonitor monitor) {
         }
     }
 
-    private void removePreviousMarkers(SubMonitor progress) throws CoreException {
-        Collection<String> previouslyKnownDerivedResources = getKnownItems();
+    private void markBuildFolder(Optional<IPath> buildDirectoryPath) throws CoreException {
+        PersistentModel preferences = CorePlugin.modelPersistence().loadModel(this.project);
+        preferences.setBuildDir(buildDirectoryPath);
+        CorePlugin.modelPersistence().saveModel(preferences);
+    }
+
+    private List<IResource> getDerivedResources(Optional<IPath> possibleBuildDirectoryPath, SubMonitor progress) {
+        List<IResource> derivedResources = Lists.newArrayList();
+        IResource dotGradle = this.project.findMember(".gradle");
```

Comment:
We shouldn't work with `IResource` here, but with `IPath`. The `.gradle` folder should be in the list of derived resources no matter if it currently exists or not. 

Then we can add a resource change listener which listens for new resources and marks them as derived if they are in the derived list.